### PR TITLE
LLM reasoning with tables (using TabMWP)

### PR DESCRIPTION
**Pull request by redoctopus**
First draft merging new LLM code with edits to get SFT, eval, etc. of tabular data working.

Most notably, added few shot examples and tables context options, plus a few minor changes to get eval/synthetic data generation/SFT to run.

### CODE REVIEW

File: datasets/tabmwp/prepare.py [hunk 1]
```diff
@@ -12,6 +12,7 @@
 # See the License for the specific language governing permissions and
```
Comment (Kipok) [inline datasets/tabmwp/prepare.py:12]: Please add more calls to this script inside dataseta/prepare_all.sh as I assume there are some subsets other than test that need to be loaded?

File: nemo_skills/inference/prompt/few_shot_examples/examples_tabmwp.py [hunk 2]
```diff
@@ -0,0 +1,217 @@
...
+$275 | 20,500 | 9,400
+$395 | 18,400 | 13,000
+$515 | 16,300 | 16,600
+$635 | 14,200 | 20,200
+""".strip(),
+        "reference_solution": "At the price of $155, the quantity demanded is greater than the quantity supplied. There is not enough of the good or service for sale at that price. So, there is a shortage.\n",
+        "generated_solution": """
+Let's solve this problem using Python code and the values from the table.
+{start_code}
+quantity_demanded_price_155 = 22600
```
Comment (Kipok) [inline nemo_skills/inference/prompt/few_shot_examples/examples_tabmwp.py:61]: Note that you don't have to use code if the problem does not require it. In this case it looks like a more natural solution would be to just solve it in text

File: nemo_skills/inference/prompt/few_shot_examples/examples_tabmwp.py [hunk 3]
```diff
@@ -0,0 +1,217 @@
...
+{start_code}
+cans = [7, 4, 5, 8, 9]
+cans = sorted(cans)
+middle1 = (len(cans) - 1) // 2
+middle2 = len(cans) // 2
+result = (cans[middle1] + cans[middle2]) / 2
+result
+{end_code}
+{start_code_output}
+7
```
Comment (Kipok) [inline nemo_skills/inference/prompt/few_shot_examples/examples_tabmwp.py:98]: I think the actual output that Python gives is "7.0". Suggest you double check it in all places to make sure the few-shot examples are more representative of what the model will see during real inference

File: nemo_skills/inference/prompt/few_shot_examples/examples_tabmwp.py [hunk 4]
```diff
@@ -0,0 +1,217 @@
...
+{start_code_output}
+yes
+{end_code_output}
+\\boxed{Yes}, Lorenzo does have enough money to buy a toy guitar and juggling balls.
+""".strip(),
+    },
+]
+
+
+only_code = [
```
Comment (Kipok) [inline nemo_skills/inference/prompt/few_shot_examples/examples_tabmwp.py:137]: I think this will not work in the current code version - have you used this few-shots? If not, maybe remove them

File: pipeline/prepare_eval.py [hunk 5]
```diff
@@ -64,6 +64,7 @@
     args.output_path.mkdir(exist_ok=True, parents=True)
 
     conversion_step = f"&& mv /training_folder/model-averaged.nemo /inference_folder/{model_name}"
+    conversion_step += f" && cp /nemo_model/model_config.yaml /inference_folder/"
```
Comment (Kipok) [inline pipeline/prepare_eval.py:67]: just wondering why is this needed?

File: pipeline/run_eval.py [hunk 6]
```diff
@@ -98,6 +99,7 @@ def get_sampling_cmd(benchmark, random_seed, extra_eval_args="", extra_arguments
     wrapper_args.add_argument("--model_path", required=True)
     wrapper_args.add_argument("--server_type", choices=('nemo', 'tensorrt_llm'), default='tensorrt_llm')
     wrapper_args.add_argument("--output_dir", required=True)
+    wrapper_args.add_argument("--context_type", default="empty")
```
Comment (Kipok) [inline pipeline/run_eval.py:102]: You don't need to add an extra parameter here as you can always add `++prompt.context_type=table` to the end of the command and it will work. So let's remove it to not have redundant options

Comment (Kipok) [reply pipeline/run_eval.py:102]: (and put empty as the default back in the cmd above)

File: pipeline/run_eval.py [hunk 7]
```diff
@@ -164,7 +167,11 @@ def get_sampling_cmd(benchmark, random_seed, extra_eval_args="", extra_arguments
     eval_cmds = [" ".join(eval_cmds[i :: args.num_nodes]) for i in range(args.num_nodes)]
 
     for idx, eval_cmd in enumerate(eval_cmds):
-        extra_sbatch_args = ["--parsable", f"--output={args.output_dir}/slurm_logs_eval{idx}.log"]
+        extra_sbatch_args = [
+            "--parsable",
+            f"--output={args.output_dir}/slurm_logs_eval{idx}.log",
+            f"--nv-meta ml-model.llm-reasoning ",
```
Comment (Kipok) [inline pipeline/run_eval.py:173]: This should go as part of the extra_sbatch_args to cluster config and not in the code

File: pipeline/run_labeling.py [hunk 8]
```diff
@@ -62,7 +62,7 @@ def run_script(format_dict, seed, extra_arguments, partition=None, dependency=No
     format_dict["random_seed"] = seed
     format_dict["extra_arguments"] = extra_arguments
 
-    extra_sbatch_args = ["--parsable", f"--output={LOGS.format(**format_dict)}"]
+    extra_sbatch_args = ["--parsable", f"--output={LOGS.format(**format_dict)}", f"--nv-meta ml-model.llm-reasoning "]
```
Comment (Kipok) [inline pipeline/run_labeling.py:65]: This should go as part of the extra_sbatch_args to cluster config and not in the code

File: pipeline/run_sft_and_eval.py [hunk 9]
```diff
@@ -27,11 +27,10 @@
 
 def start_job(extra_sbatch_args: str, cmd: str) -> int:
     full_cmd = f'{extra_sbatch_args} {cmd}'
+
     if CLUSTER_CONFIG['cluster'] == 'local':
         subprocess.run(full_cmd, shell=True)
         return 0
-    job_id = subprocess.run(full_cmd, shell=True, check=True, capture_output=True).stdout.decode()
```
Comment (Kipok) [inline pipeline/run_sft_and_eval.py:33]: Accidentally removed?

**Review by Kipok — CHANGES_REQUESTED** [review]

File: pipeline/run_sft_and_eval.py (hunk 9 above)
Comment (redoctopus) [reply pipeline/run_sft_and_eval.py:33]: Yep, thanks for catching this!

File: nemo_skills/inference/prompt/few_shot_examples/examples_tabmwp.py (hunk 3 above)
Comment (redoctopus) [reply nemo_skills/inference/prompt/few_shot_examples/examples_tabmwp.py]: Oh, nice catch, thanks. Just ran all the example solutions to double-check the outputs and found a few more issues.

File: nemo_skills/inference/prompt/few_shot_examples/examples_tabmwp.py (hunk 2 above)
Comment (redoctopus) [reply nemo_skills/inference/prompt/few_shot_examples/examples_tabmwp.py:61]: Do we have a way of mixing code/text solutions with the model being able to choose what to use?

In this case the end goal is to have the values loaded programmatically so I figure these examples should all involve translating the problem to code to begin with.

File: nemo_skills/inference/prompt/few_shot_examples/examples_tabmwp.py (hunk 4 above)
Comment (redoctopus) [reply nemo_skills/inference/prompt/few_shot_examples/examples_tabmwp.py:137]: Odd, it seemed to be working for me a few weeks back--I used it to make prompts for synthetic data generation. When I checked there was definitely a difference between the results from `only_code` vs. `text_with_code`.

What is expected behavior for the current version?

File: pipeline/run_eval.py (hunk 6 above)
Comment (redoctopus) [reply pipeline/run_eval.py:102]: Ah, I was originally worried that having it specified twice in the command would cause an error.

Just to make sure--will the `++prompt.context_type=table` additional arg overwrite the existing `prompt.context_type=empty` if both are specified?

File: nemo_skills/inference/prompt/few_shot_examples/examples_tabmwp.py (hunk 2 above)
Comment (Kipok) [reply nemo_skills/inference/prompt/few_shot_examples/examples_tabmwp.py:61]: The model is supposed to learn that. So if you generate some examples that use code and some that don't and train on those, the hope is that it will be able to differentiate between these two cases at inference time. But you need to be showing this in few shots (either in the same set of few shots show both or have example with/without code and run multiple labeling jobs)

File: nemo_skills/inference/prompt/few_shot_examples/examples_tabmwp.py (hunk 4 above)
Comment (Kipok) [reply nemo_skills/inference/prompt/few_shot_examples/examples_tabmwp.py:137]: the code you generate is not going to be run unless you wrap it with the special tokens. Since there are no such tokens here, the model will still learn to produce the code, but will never see its output

File: pipeline/run_eval.py (hunk 6 above)
Comment (Kipok) [reply pipeline/run_eval.py:102]: yes, it will overwrite

File: nemo_skills/inference/prompt/few_shot_examples/examples_tabmwp.py (hunk 2 above)
Comment (redoctopus) [reply nemo_skills/inference/prompt/few_shot_examples/examples_tabmwp.py:61]: Gotcha, makes sense. I'll keep these with code examples for now since we're focusing on code-generation, but good to keep in mind.

File: pipeline/prepare_eval.py (hunk 5 above)
Comment (redoctopus) [reply pipeline/prepare_eval.py:67]: Some part of the eval process where the model was being restored wanted this file in the same directory as the checkpoint, but I can't remember exactly where it was that was erroring out without this...

File: nemo_skills/inference/prompt/few_shot_examples/examples_tabmwp.py (hunk 4 above)
Comment (redoctopus) [reply nemo_skills/inference/prompt/few_shot_examples/examples_tabmwp.py:137]: Yeah, that checks out. I don't think we use this format so I'll remove these examples.

File: pipeline/prepare_eval.py (hunk 5 above)
Comment (Kipok) [reply pipeline/prepare_eval.py:67]: It shouldn't be needed, so let's remove it for now. If you get an error again, please let me know

File: pipeline/run_labeling.py (hunk 8 above)
Comment (redoctopus) [reply pipeline/run_labeling.py]: Done

File: pipeline/run_eval.py (hunk 7 above)
Comment (redoctopus) [reply pipeline/run_eval.py]: Done

**Review by Kipok — APPROVED** [review]
looks good, thanks!

**Comment by Kipok** [comment]
Actually looks like one of the tests is failing because something is wrong with dataset creation. Can you please fix that? Also add other splits into the test file so that we check that they are created as well

**Comment by redoctopus** [comment]
> Actually looks like one of the tests is failing because something is wrong with dataset creation. Can you please fix that? Also add other splits into the test file so that we check that they are created as well

Ah whoops, looks like I was saving them as .json instead of the expected .jsonl. Fixed.
